filecoin-proofs/src/api.rs

```diff
@@ -263,6 +373,51 @@ pub fn verify_seal(
     .map_err(Into::into)
 }
 
+/// Verify the provided piece inclusion proof demonstrates the piece commitment exists in a
+/// merkle tree of a specific size with root hash comm_d
+pub fn verify_piece_inclusion_proof(
+    piece_inclusion_proof: &[u8],
+    comm_d: &Commitment,
+    comm_p: &Commitment,
+    piece_size: PaddedBytesAmount,
+    sector_size: SectorSize,
+) -> error::Result<bool> {
+    let piece_inclusion_proof: PieceInclusionProof<PedersenHasher> =
+        piece_inclusion_proof.try_into()?;
+    let comm_d = storage_proofs::hasher::pedersen::PedersenDomain::try_from_bytes(comm_d)?;
+    let comm_p = storage_proofs::hasher::pedersen::PedersenDomain::try_from_bytes(comm_p)?;
+    let piece_leaves = u64::from(piece_size) / 32;
```

Comment:
Is `piece_size` guaranteed to be a multiple of 32?

Reply:
Yes. 

The `piece_size` here (perhaps unfortunately named) is the value produced by the `generate_piece_commitment` call. This value represents the number of bytes that the piece occupies in the sealed sector, which is equal to the size of the piece as provided by the user PLUS bit-padding PLUS alignment bytes. Alignment bytes is a function of the piece's position in the sector; simply having the number of user-provided bytes in the piece is not enough for a verifier to know how many bytes of alignment were added.

Reply:
@sidke - Perhaps renaming to `piece_size_with_alignment` would help to disambiguate.

Reply:
Update:

> Alignment bytes is a function of the piece's position in the sector; simply having the number of user-provided bytes in the piece is not enough for a verifier to know how many bytes of alignment were added.

The above statement (from me) is false. Alignment bytes are not a function of a piece's position in the sector. In as much as we must left-pad a piece in order to achieve the correct alignment, those 0s are not included in the CommP computation.

We have updated the changeset:

1. `generate_piece_commitment` accepts a piece file and the number of plaintext user-bytes to read from that file
1. `verify_piece_inclusion_proof` accepts the number of bytes in the original file and computes alignment on the fly